src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -5305,6 +5305,21 @@ MergeTreeData::DataPartsVector MergeTreeData::getAllDataPartsVector(MergeTreeDat
     return res;
 }
 
+size_t MergeTreeData::getAllPartsCount() const
+{
+    return data_parts_by_info.size();
```

Comment:
I'm not sure that it is safe to call .size() without locking data_parts_mutex (lockParts). It is definitely concurrently modifying by others threads.

Reply:
I think the best way to do this is like https://github.com/ClickHouse/ClickHouse/pull/17639, add two async metrics to record both total parts count and total marks count, such that we don't need to lock parts anymore.

Reply:
That approach in #17639 is even better. But it is also harder. Designated counter should be managed with respect to all modifications of data_parts_by_info.

I'm don't mind if it uses locks. Like this
```
size_t MergeTreeData::getAllPartsCount() const
{
    auto lock = lockParts();
    return data_parts_by_info.size();
}
```